### PR TITLE
MOL-354: Improved Order State Handling

### PR DESCRIPTION
**Pull request by ThLind**
Improved Order State Handling. In the old state handling unnecessary and failing transitions were executed to bring the order in known states. Now only perform necessary transitions and avoid failing transitions.

### CODE REVIEW

File: src/Helper/OrderTransitionHelper.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+use Shopware\Core\Checkout\Order\OrderDefinition;
+use Shopware\Core\Checkout\Order\OrderEntity;
+use Shopware\Core\Checkout\Order\OrderStates;
+use Shopware\Core\Framework\Context;
+use Shopware\Core\System\StateMachine\Aggregation\StateMachineTransition\StateMachineTransitionActions;
+use Shopware\Core\System\StateMachine\Aggregation\StateMachineTransition\StateMachineTransitionEntity;
+use Shopware\Core\System\StateMachine\StateMachineRegistry;
+use Shopware\Core\System\StateMachine\Transition;
+
+class OrderTransitionHelper implements OrderTransitionHelperInterface
```
Comment (schmitzcarsten) [inline src/Helper/OrderTransitionHelper.php:16]: Please don't use class names like helper. Find a meaningful service name like TransitionOrderStatusService or something like that. Helper could be everything and nothing.

File: src/Helper/OrderStateHelper.php [hunk 2]
```diff
@@ -5,29 +5,25 @@
...
 use Kiener\MolliePayments\Setting\MollieSettingStruct;
-use Shopware\Core\Checkout\Order\OrderDefinition;
 use Shopware\Core\Checkout\Order\OrderEntity;
 use Shopware\Core\Checkout\Order\OrderStates;
 use Shopware\Core\Framework\Context;
-use Shopware\Core\System\StateMachine\Aggregation\StateMachineTransition\StateMachineTransitionActions;
-use Shopware\Core\System\StateMachine\StateMachineRegistry;
-use Shopware\Core\System\StateMachine\Transition;
 
 class OrderStateHelper
```
Comment (schmitzcarsten) [inline src/Helper/OrderStateHelper.php:12]: Please don't use class names like helper. Find a meaningful service name like ChangeOrderState or something like that. Helper could be everything and nothing.

File: src/Helper/OrderTransitionHelperInterface.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
+<?php declare(strict_types=1);
+
+
+namespace Kiener\MolliePayments\Helper;
+
+
+use Shopware\Core\Checkout\Order\OrderEntity;
+use Shopware\Core\Framework\Context;
+
+interface OrderTransitionHelperInterface
```
Comment (schmitzcarsten) [inline src/Helper/OrderTransitionHelperInterface.php:10]: find a better name

File: tests/Fakes/FakeOrderTransitionHelper.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+namespace Kiener\MolliePayments\Tests\Fakes;
+
+
+use Exception;
+use Kiener\MolliePayments\Helper\OrderTransitionHelperInterface;
+use Shopware\Core\Checkout\Order\OrderEntity;
+use Shopware\Core\Framework\Context;
+
+class FakeOrderTransitionHelper implements OrderTransitionHelperInterface
```
Comment (schmitzcarsten) [inline tests/Fakes/FakeOrderTransitionHelper.php:11]: rename accordingly to the class that is tested

File: tests/Helper/OrderStateHelperTest.php [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+use Kiener\MolliePayments\Tests\Fakes\FakeOrderTransitionHelper;
+use PHPUnit\Framework\TestCase;
+use Shopware\Core\Checkout\Order\OrderEntity;
+use Shopware\Core\Checkout\Order\OrderStates;
+use Shopware\Core\Framework\Context;
+use Shopware\Core\Framework\DataAbstractionLayer\Event\EntityWrittenContainerEvent;
+use Shopware\Core\Framework\Log\LogEntryDefinition;
+use Shopware\Core\System\StateMachine\Aggregation\StateMachineState\StateMachineStateEntity;
+
+class OrderStateHelperTest extends TestCase
```
Comment (schmitzcarsten) [inline tests/Helper/OrderStateHelperTest.php:19]: rename accordingly to the class that is tested

File: tests/Helper/OrderTransitionHelperTest.php [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+use Shopware\Core\Checkout\Order\OrderStates;
+use Shopware\Core\Framework\Context;
+use Shopware\Core\Framework\Log\LogEntryDefinition;
+use Shopware\Core\System\StateMachine\Aggregation\StateMachineState\StateMachineStateEntity;
+use Shopware\Core\System\StateMachine\Aggregation\StateMachineTransition\StateMachineTransitionActions;
+use Shopware\Core\System\StateMachine\Aggregation\StateMachineTransition\StateMachineTransitionEntity;
+use Shopware\Core\System\StateMachine\StateMachineRegistry;
+use Shopware\Core\System\StateMachine\Transition;
+
+class OrderTransitionHelperTest extends TestCase
```
Comment (schmitzcarsten) [inline tests/Helper/OrderTransitionHelperTest.php:21]: rename accordingly to the class that is tested

**Review by schmitzcarsten — CHANGES_REQUESTED** [review]

File: src/Helper/OrderStateHelper.php (hunk 2 above)
Comment (ThLind) [reply src/Helper/OrderStateHelper.php:12]: Thank the earlier naming for this ;-) Just didn't want to change this everywhere it's used and leave this for your big refactoring, but if you want I can do it the next time I do work on this.

Comment (schmitzcarsten) [reply src/Helper/OrderStateHelper.php:12]: yes please. Move it to a suitable namespace and rename it. 
But only such classes that you have added or updated. Then it should not be much work

**Review by schmitzcarsten — APPROVED** [review]

**Review by barbieswimcrew — APPROVED** [review]

**Pull request merged**